### PR TITLE
#97, #102, #103, #109, #112 - implemented conversion between abi encoded entities and swift entities

### PR DESCRIPTION
**Pull request by biboran**
As per #97 
* Added decoding for all abi types except for `int` and dynamic collection of static collections (i.e. tuples).
* Added `BoolScalar`, `CollectionScalar` and some relevant utils to handle the decoding. 

As per #102 
* Offset is no longer computed from the beginning of the whole encoding and is now computed from the relevant position in the dynamic collection.

As per #103 
* Rebased the swiftlint fork, updated it and added exception to the problematic rule 

As per #109 
* Removed `colon` rule for now due to realm/SwiftLint/issues/1901
* Removed `function_body_length` because it's not relevant in tests
* Updated some outdated dependencies 

As per #112 
* Set a build flag `ONLY_ACTIVE_ARCH=YES`

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#110](https://github.com/blockstoreapp/web3swift/issues/110) is now in scope, role is `REV`

File: Web3Swift/CollectionScalar/SizeConstrainedCollection.swift [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import Foundation
+
+internal final class IndexOtOfBoundsError: DescribedError {
```
[the remark this answers is not in the document]
Comment (biboran) [reply Web3Swift/CollectionScalar/SizeConstrainedCollection.swift]: @kondranton right, thanks!

File: Web3Swift/ABI/Decoding/DecodedABIDynamicCollection.swift [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+limitations under the License.
+*/
+
+import Foundation
+
+//Decoded typed abi collection
+public final class DecodedABIDynamicCollection<T>: CollectionScalar<T> {
+
+    private let abiMessage: CollectionScalar<BytesScalar>
+    private let mapping: ((slice: CollectionScalar<BytesScalar>, index: UInt)) throws -> (T)
```
[the remark this answers is not in the document]
Comment (biboran) [reply Web3Swift/ABI/Decoding/DecodedABIDynamicCollection.swift]: @driver733 could you please give a reason for such rule?

[the remark this answers is not in the document]
Comment (driver733) [reply Web3Swift/ABI/Decoding/DecodedABIDynamicCollection.swift]: > The motivation for such formatting is simple - we want to see the entire block as fast as possible. When you look at a block of code you should be able to see where it starts and where it ends. In exactly the same way we organize curled brackets.
 In other words, when you open a bracket and can't close it at the same line - you should leave it as the last symbol at this line.

File: Example/Tests/ABI/Decoding/ABIDecodingIT.swift [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+                index: 2
+            ).uint()
+        }.to(
+            equal(arguments.2),
+            description: "Encoded argument \(arguments.2) is expected to persist"
+        )
+    }
+
+    //address[], bool, string[]
+    func testDynamicCollectionOfAddressesBoolDynamicCollectionOfStringsIsDecodedCorrectly() {
```
[the remark this answers is not in the document]
Comment (biboran) [reply Example/Tests/ABI/Decoding/ABIDecodingIT.swift]: @driver733 do you have any idea how to cluster such tests or how to call them reasonably?

File: Example/Tests/TransactionBytes/ContractDeploymentAndCallIT.swift [hunk 4]
```diff
@@ -183,6 +183,124 @@ final class ContractDeploymentAndCallIT: XCTestCase {
...
...56103d75783830388526103b9838351610443565b92506103c48261057e565b91506020880197506001810190506103a1565b508196508694505050505092915050565b60006103f382610573565b80845260208401935061040583610550565b60005b828110156104375761041b86835161049e565b61042482610598565b9150602086019550600181019050610408565b50849250505092915050565b600061044e82610568565b80845260208401935061046083610543565b60005b828110156104925761047686835161049e565b61047f8261058b565b9150602086019550600181019050610463565b50849250505092915050565b6104a7816105a5565b82525050565b6104b6816105af565b82525050565b600060208201905081810360008301526104d6818461037a565b905092915050565b600060208201905081810360008301526104f881846103e8565b905092915050565b6000602082019050610515600083018461049e565b92915050565b600060208201905061053060008301846104ad565b92915050565b6000602082019050919050565b6000602082019050919050565b6000602082019050919050565b600081519050919050565b600081519050919050565b600081519050919050565b6000602082019050919050565b6000602082019050919050565b6000602082019050919050565b6000819050919050565b6000819050919050565b60008190509190505600a265627a7a72305820e4eba7dab74878c23493b48feb97733749dbd3d9a238fbdaee586052f5ab435d6c6578706572696d656e74616cf50037"
```
[the remark this answers is not in the document]
Comment (biboran) [reply Example/Tests/TransactionBytes/ContractDeploymentAndCallIT.swift]: @driver733 this one is ok

**Comment by biboran** [comment]
@kondranton @driver733 ping

**Pull request merged**